src/app/reporting/LoggingManagement.h

```diff
@@ -0,0 +1,571 @@
+     *
+     * @param[in] apPrev         The pointer to CircularEventBuffer storing
+     *                           events of lesser priority.
+     *
+     * @param[in] apNext         The pointer to CircularEventBuffer storing
+     *                           events of greater priority.
+     *
+     * @return CircularEventBuffer
+     */
+    CircularEventBuffer(uint8_t * apBuffer, uint32_t aBufferLength, CircularEventBuffer * apPrev, CircularEventBuffer * apNext);
+
+    CHIP_ERROR GetNextBuffer(chip::TLV::TLVReader & aReader, const uint8_t *& aBufStart, uint32_t & aBufLen) override;
+
+    /**
+     * @brief
+     *   A helper function that determines whether the event of
+     *   specified priority is dropped from this buffer.
+     *
+     * @param[in]   aPriority   Priority of the event.
+     *
+     * @retval true  The event will be dropped from this buffer as
+     *               a result of queue overflow.
+     * @retval false The event will be bumped to the next queue.
+     */
+    bool IsFinalDestinationForPriority(PriorityLevel aPriority) const;
+
+    /**
+     * @brief
+     *   Allocate a new event ID based on the event priority, and advance the counter
+     *   if we have one.
+     *
+     * @return chip::EventNumber Event Number for this priority.
+     */
+    chip::EventNumber VendEventNumber(void);
+    void RemoveEvent(uint16_t aNumEvents);
+
+    /**
+     * @brief
+     *   Given a timestamp of an event, compute the delta time to store in the log
+     *
+     * @param aEventTimestamp The event timestamp.
+     *
+     * @return int32_t         Time delta to encode for the event.
+     */
+    void AddEvent(timestamp_t aEventTimestamp);
+
+private:
+    friend class LoggingManagement;
+    friend class CircularEventReader;
+
+    CircularEventBuffer * mpPrev; //< A pointer CircularEventBuffer storing events less important events
+    CircularEventBuffer * mpNext; //< A pointer CircularEventBuffer storing events more important events
+
+    PriorityLevel mPriority; //< The buffer is the final bucket for events of this priority.  Events of lesser priority are
+                             //< dropped when they get bumped out of this buffer
+
+    chip::EventNumber mFirstEventNumber; //< First event Number stored in the logging subsystem for this priority
+    chip::EventNumber mLastEventNumber;  //< Last event Number vended for this priority
+
+    timestamp_t mFirstEventTimestamp; //< The timestamp of the first event in this buffer
+    timestamp_t mLastEventTimestamp;  //< The timestamp of the last event in this buffer
+
+    // The counter we're going to actually use.
+    chip::MonotonicallyIncreasingCounter * mpEventNumberCounter;
+
+    // The backup counter to use if no counter is provided for us.
+    chip::MonotonicallyIncreasingCounter mNonPersistedCounter;
+};
+
+/**
+ * @brief
+ *   A TLVReader backed by CircularEventBuffer
+ */
+class CircularEventReader : public chip::TLV::TLVReader
```

Comment:
Same thing: does this need to be in the header?

Reply:
thanks, just moved it to cpp file